### PR TITLE
Fix for bundler master

### PR DESCRIPTION
**Pull request by segiddins**
Adds swapping when a conflict occurs

\c @alloy @indirect @Who828 @floere

### CODE REVIEW

**Comment by segiddins** [comment]
For everyone who's interested, my object to merging this is that I want the integration specs to be entirely semver compliant, to make it easier for other projects to use these specs.

The reason I changed to `Gem::Version` in the specs is merely one of convenience -- that rubygems failing spec has nothing to do with the versioning scheme, but rather a desire to be able to quickly translate issues with Bundler into proper Molinillo test cases. In short, this change is a matter of _my own laziness_. Gem dependency graphs have a tendency to get massive, and manually converting the Rubygems index to be semver-compliant will take a _lot_ of time.

**Comment by alloy** [comment]
@segiddins Thanks for that clarification :+1:

**Comment by segiddins** [comment]
OK, managed to come up with a script to 'properly' sanitize the rubygems index to be SemVer compliant -- please review the actual changes to the resolver so we can merge.

**Comment by alloy** [comment]
Can you please explain what it means to "'properly' sanitize the rubygems index to be SemVer compliant"? Are there any repercussions, i.e. does this require a change to how RubyGems works?

**Comment by segiddins** [comment]
No, it just means some versions are dropped, but I've done it in such a way that I think I'll easily be able to port over gem-based cases to integration specs. 

-Samuel E. Giddins

On Feb 8, 2015, at 1:47 AM, Eloy Durán <notifications@github.com> wrote:

Can you please explain what it means to "'properly' sanitize the rubygems index to be SemVer compliant"? Are there any repercussions, i.e. does this require a change to how RubyGems works?

**Comment by segiddins** [comment]
@Who828 please review?

File: lib/molinillo/resolution.rb [hunk 1]
```diff
@@ -320,12 +320,24 @@ def attempt_to_activate_existing_spec(existing_node)
...
         end
       end
 
+      # Attempts to swp the current {#possibility} with the already-activated
+      # spec with the given name
+      # @return [Boolean] Whether the possibility was swapped into {#activated}
+      def attempt_to_swap_possibility
+        swapped = activated.dup
+        swapped.vertex_named(name).payload = possibility
+        return unless swapped.vertex_named(name).requirements.
```
Comment (kylef) [inline lib/molinillo/resolution.rb]: The method documentation states this returns boolean but isn't the case here?

Comment (segiddins) [inline lib/molinillo/resolution.rb]: Nil is a Boolean value in Ruby 

-Samuel E. Giddins

On Feb 14, 2015, at 7:38 AM, Kyle Fuller <notifications@github.com> wrote:

In lib/molinillo/resolution.rb:

>            create_conflict
>            debug(depth) { "Unsatisfied by existing spec (#{existing_node.payload})" }
>            unwind_for_conflict
>          end
>        end
>  
> +      # Attempts to swp the current {#possibility} with the already-activated
> +      # spec with the given name
> +      # @return [Boolean] Whether the possibility was swapped into {#activated}
> +      def attempt_to_swap_possibility
> +        swapped = activated.dup
> +        swapped.vertex_named(name).payload = possibility
> +        return unless swapped.vertex_named(name).requirements.
The method documentation states this returns boolean but isn't the case here?

**Comment by segiddins** [comment]
Merging on 2/20 unless anyone raises any objections.

**Pull request merged**